In [2]:
import sqlite3
import pandas as pd

In [3]:
# creating database connection
conn = sqlite3.connect('inventory.db')

In [4]:
# checking tables present in the database
tables = pd.read_sql_query("SELECT name FROM sqlite_master WHERE type='table'",conn)
tables

name
0  begin_inventory
1    end_inventory
2        purchases
3  purchase_prices
4            sales
5   vendor_invoice

In [5]:
# exploring what type of data available in the tables
for table in tables['name']:
    print('-'*50,f'{table}','-'*50)
    print('Count of records:',pd.read_sql_query(f"select count(*) as count from {table}",conn)['count'].values[0])
    display(pd.read_sql_query(f"select * from {table} limit 5",conn))
    

-------------------------------------------------- begin_inventory --------------------------------------------------
Count of records: 206529


InventoryId  Store          City  Brand                  Description  \
0  1_HARDERSFIELD_58      1  HARDERSFIELD     58  Gekkeikan Black & Gold Sake   
1  1_HARDERSFIELD_60      1  HARDERSFIELD     60       Canadian Club 1858 VAP   
2  1_HARDERSFIELD_62      1  HARDERSFIELD     62     Herradura Silver Tequila   
3  1_HARDERSFIELD_63      1  HARDERSFIELD     63   Herradura Reposado Tequila   
4  1_HARDERSFIELD_72      1  HARDERSFIELD     72         No. 3 London Dry Gin   

    Size  onHand  Price   startDate  
0  750mL       8  12.99  2024-01-01  
1  750mL       7  10.99  2024-01-01  
2  750mL       6  36.99  2024-01-01  
3  750mL       3  38.99  2024-01-01  
4  750mL       6  34.99  2024-01-01

-------------------------------------------------- end_inventory --------------------------------------------------
Count of records: 224489


InventoryId  Store          City  Brand                  Description  \
0  1_HARDERSFIELD_58      1  HARDERSFIELD     58  Gekkeikan Black & Gold Sake   
1  1_HARDERSFIELD_62      1  HARDERSFIELD     62     Herradura Silver Tequila   
2  1_HARDERSFIELD_63      1  HARDERSFIELD     63   Herradura Reposado Tequila   
3  1_HARDERSFIELD_72      1  HARDERSFIELD     72         No. 3 London Dry Gin   
4  1_HARDERSFIELD_75      1  HARDERSFIELD     75    Three Olives Tomato Vodka   

    Size  onHand  Price     endDate  
0  750mL      11  12.99  2024-12-31  
1  750mL       7  36.99  2024-12-31  
2  750mL       7  38.99  2024-12-31  
3  750mL       4  34.99  2024-12-31  
4  750mL       7  14.99  2024-12-31

-------------------------------------------------- purchases --------------------------------------------------
Count of records: 2372474


InventoryId  Store  Brand                   Description   Size  \
0    69_MOUNTMEND_8412     69   8412     Tequila Ocho Plata Fresno  750mL   
1     30_CULCHETH_5255     30   5255  TGI Fridays Ultimte Mudslide  1.75L   
2    34_PITMERDEN_5215     34   5215  TGI Fridays Long Island Iced  1.75L   
3  1_HARDERSFIELD_5255      1   5255  TGI Fridays Ultimte Mudslide  1.75L   
4    76_DONCASTER_2034     76   2034     Glendalough Double Barrel  750mL   

   VendorNumber                   VendorName  PONumber      PODate  \
0           105  ALTAMAR BRANDS LLC               8124  2023-12-21   
1          4466  AMERICAN VINTAGE BEVERAGE        8137  2023-12-22   
2          4466  AMERICAN VINTAGE BEVERAGE        8137  2023-12-22   
3          4466  AMERICAN VINTAGE BEVERAGE        8137  2023-12-22   
4           388  ATLANTIC IMPORTING COMPANY       8169  2023-12-24   

  ReceivingDate InvoiceDate     PayDate  PurchasePrice  Quantity  Dollars  \
0    2024-01-02  2024-01-04  2024-02-16          35.71         6   214.26   
1    2024-01-01  2024-01-07  2024-02-21           9.35         4    37.40   
2    2024-01-02  2024-01-07  2024-02-21           9.41         5    47.05   
3    2024-01-01  2024-01-07  2024-02-21           9.35         6    56.10   
4    2024-01-02  2024-01-09  2024-02-16          21.32         5   106.60   

   Classification  
0               1  
1               1  
2               1  
3               1  
4               1

-------------------------------------------------- purchase_prices --------------------------------------------------
Count of records: 12261


Brand                  Description  Price   Size Volume  Classification  \
0     58  Gekkeikan Black & Gold Sake  12.99  750mL    750               1   
1     62     Herradura Silver Tequila  36.99  750mL    750               1   
2     63   Herradura Reposado Tequila  38.99  750mL    750               1   
3     72         No. 3 London Dry Gin  34.99  750mL    750               1   
4     75    Three Olives Tomato Vodka  14.99  750mL    750               1   

   PurchasePrice  VendorNumber                   VendorName  
0           9.28          8320  SHAW ROSS INT L IMP LTD      
1          28.67          1128  BROWN-FORMAN CORP            
2          30.46          1128  BROWN-FORMAN CORP            
3          26.11          9165  ULTRA BEVERAGE COMPANY LLP   
4          10.94          7245  PROXIMO SPIRITS INC.

-------------------------------------------------- sales --------------------------------------------------
Count of records: 12825363


InventoryId  Store  Brand                 Description        Size  \
0  1_HARDERSFIELD_1004      1   1004  Jim Beam w/2 Rocks Glasses       750mL   
1  1_HARDERSFIELD_1004      1   1004  Jim Beam w/2 Rocks Glasses       750mL   
2  1_HARDERSFIELD_1004      1   1004  Jim Beam w/2 Rocks Glasses       750mL   
3  1_HARDERSFIELD_1004      1   1004  Jim Beam w/2 Rocks Glasses       750mL   
4  1_HARDERSFIELD_1005      1   1005     Maker's Mark Combo Pack  375mL 2 Pk   

   SalesQuantity  SalesDollars  SalesPrice   SalesDate  Volume  \
0              1         16.49       16.49  2024-01-01   750.0   
1              2         32.98       16.49  2024-01-02   750.0   
2              1         16.49       16.49  2024-01-03   750.0   
3              1         14.49       14.49  2024-01-08   750.0   
4              2         69.98       34.99  2024-01-09   375.0   

   Classification  ExciseTax  VendorNo                   VendorName  
0               1       0.79     12546  JIM BEAM BRANDS COMPANY      
1               1       1.57     12546  JIM BEAM BRANDS COMPANY      
2               1       0.79     12546  JIM BEAM BRANDS COMPANY      
3               1       0.79     12546  JIM BEAM BRANDS COMPANY      
4               1       0.79     12546  JIM BEAM BRANDS COMPANY

-------------------------------------------------- vendor_invoice --------------------------------------------------
Count of records: 5543


VendorNumber                   VendorName InvoiceDate  PONumber  \
0           105  ALTAMAR BRANDS LLC           2024-01-04      8124   
1          4466  AMERICAN VINTAGE BEVERAGE    2024-01-07      8137   
2           388  ATLANTIC IMPORTING COMPANY   2024-01-09      8169   
3           480  BACARDI USA INC              2024-01-12      8106   
4           516  BANFI PRODUCTS CORP          2024-01-07      8170   

       PODate     PayDate  Quantity    Dollars  Freight Approval  
0  2023-12-21  2024-02-16         6     214.26     3.47     None  
1  2023-12-22  2024-02-21        15     140.55     8.57     None  
2  2023-12-24  2024-02-16         5     106.60     4.61     None  
3  2023-12-20  2024-02-05     10100  137483.78  2935.20     None  
4  2023-12-24  2024-02-12      1935   15527.25   429.20     None

In [6]:
# choosing one vendor and filtering the different tables 
purchases = pd.read_sql_query("select * from purchases where VendorNumber = 4466",conn)
purchases

InventoryId  Store  Brand                   Description   Size  \
0        30_CULCHETH_5255     30   5255  TGI Fridays Ultimte Mudslide  1.75L   
1       34_PITMERDEN_5215     34   5215  TGI Fridays Long Island Iced  1.75L   
2     1_HARDERSFIELD_5255      1   5255  TGI Fridays Ultimte Mudslide  1.75L   
3       38_GOULCREST_5215     38   5215  TGI Fridays Long Island Iced  1.75L   
4     59_CLAETHORPES_5215     59   5215  TGI Fridays Long Island Iced  1.75L   
...                   ...    ...    ...                           ...    ...   
2187     81_PEMBROKE_5215     81   5215  TGI Fridays Long Island Iced  1.75L   
2188   62_KILMARNOCK_5255     62   5255  TGI Fridays Ultimte Mudslide  1.75L   
2189    34_PITMERDEN_5215     34   5215  TGI Fridays Long Island Iced  1.75L   
2190     6_GOULCREST_5215      6   5215  TGI Fridays Long Island Iced  1.75L   
2191    35_HALIVAARA_5255     35   5255  TGI Fridays Ultimte Mudslide  1.75L   

      VendorNumber                   VendorName  PONumber      PODate  \
0             4466  AMERICAN VINTAGE BEVERAGE        8137  2023-12-22   
1             4466  AMERICAN VINTAGE BEVERAGE        8137  2023-12-22   
2             4466  AMERICAN VINTAGE BEVERAGE        8137  2023-12-22   
3             4466  AMERICAN VINTAGE BEVERAGE        8207  2023-12-27   
4             4466  AMERICAN VINTAGE BEVERAGE        8207  2023-12-27   
...            ...                          ...       ...         ...   
2187          4466  AMERICAN VINTAGE BEVERAGE       13595  2024-12-20   
2188          4466  AMERICAN VINTAGE BEVERAGE       13595  2024-12-20   
2189          4466  AMERICAN VINTAGE BEVERAGE       13595  2024-12-20   
2190          4466  AMERICAN VINTAGE BEVERAGE       13595  2024-12-20   
2191          4466  AMERICAN VINTAGE BEVERAGE       13595  2024-12-20   

     ReceivingDate InvoiceDate     PayDate  PurchasePrice  Quantity  Dollars  \
0       2024-01-01  2024-01-07  2024-02-21           9.35         4    37.40   
1       2024-01-02  2024-01-07  2024-02-21           9.41         5    47.05   
2       2024-01-01  2024-01-07  2024-02-21           9.35         6    56.10   
3       2024-01-07  2024-01-19  2024-02-26           9.41         6    56.46   
4       2024-01-05  2024-01-19  2024-02-26           9.41         6    56.46   
...            ...         ...         ...            ...       ...      ...   
2187    2024-12-29  2025-01-04  2025-02-10           9.41         6    56.46   
2188    2024-12-28  2025-01-04  2025-02-10           9.35         5    46.75   
2189    2024-12-28  2025-01-04  2025-02-10           9.41         5    47.05   
2190    2024-12-31  2025-01-04  2025-02-10           9.41         6    56.46   
2191    2024-12-27  2025-01-04  2025-02-10           9.35         5    46.75   

      Classification  
0                  1  
1                  1  
2                  1  
3                  1  
4                  1  
...              ...  
2187               1  
2188               1  
2189               1  
2190               1  
2191               1  

[2192 rows x 16 columns]

In [7]:
purchase_prices = pd.read_sql_query("""select * from purchase_prices where VendorNumber = 4466""",conn)
purchase_prices

Brand                   Description  Price    Size Volume  Classification  \
0   5215  TGI Fridays Long Island Iced  12.99  1750mL   1750               1   
1   5255  TGI Fridays Ultimte Mudslide  12.99  1750mL   1750               1   
2   3140      TGI Fridays Orange Dream  14.99  1750mL   1750               1   

   PurchasePrice  VendorNumber                   VendorName  
0           9.41          4466  AMERICAN VINTAGE BEVERAGE    
1           9.35          4466  AMERICAN VINTAGE BEVERAGE    
2          11.19          4466  AMERICAN VINTAGE BEVERAGE

In [8]:
sales = pd.read_sql_query("""select * from sales where VendorNo = 4466""",conn)
sales

InventoryId  Store  Brand                   Description   Size  \
0     1_HARDERSFIELD_5215      1   5215  TGI Fridays Long Island Iced  1.75L   
1     1_HARDERSFIELD_5215      1   5215  TGI Fridays Long Island Iced  1.75L   
2     1_HARDERSFIELD_5215      1   5215  TGI Fridays Long Island Iced  1.75L   
3     1_HARDERSFIELD_5215      1   5215  TGI Fridays Long Island Iced  1.75L   
4     1_HARDERSFIELD_5215      1   5215  TGI Fridays Long Island Iced  1.75L   
...                   ...    ...    ...                           ...    ...   
9448     9_BLACKPOOL_5215      9   5215  TGI Fridays Long Island Iced  1.75L   
9449     9_BLACKPOOL_5255      9   5255  TGI Fridays Ultimte Mudslide  1.75L   
9450     9_BLACKPOOL_5255      9   5255  TGI Fridays Ultimte Mudslide  1.75L   
9451     9_BLACKPOOL_5255      9   5255  TGI Fridays Ultimte Mudslide  1.75L   
9452     9_BLACKPOOL_5255      9   5255  TGI Fridays Ultimte Mudslide  1.75L   

      SalesQuantity  SalesDollars  SalesPrice   SalesDate  Volume  \
0                 1         12.99       12.99  2024-01-09  1750.0   
1                 1         12.99       12.99  2024-01-12  1750.0   
2                 1         12.99       12.99  2024-01-15  1750.0   
3                 1         12.99       12.99  2024-01-21  1750.0   
4                 1         12.99       12.99  2024-01-23  1750.0   
...             ...           ...         ...         ...     ...   
9448              1         12.99       12.99  2024-12-21  1750.0   
9449              1         12.99       12.99  2024-12-02  1750.0   
9450              1         12.99       12.99  2024-12-09  1750.0   
9451              1         12.99       12.99  2024-12-23  1750.0   
9452              1         12.99       12.99  2024-12-31  1750.0   

      Classification  ExciseTax  VendorNo                   VendorName  
0                  1       1.84      4466  AMERICAN VINTAGE BEVERAGE    
1                  1       1.84      4466  AMERICAN VINTAGE BEVERAGE    
2                  1       1.84      4466  AMERICAN VINTAGE BEVERAGE    
3                  1       1.84      4466  AMERICAN VINTAGE BEVERAGE    
4                  1       1.84      4466  AMERICAN VINTAGE BEVERAGE    
...              ...        ...       ...                          ...  
9448               1       1.84      4466  AMERICAN VINTAGE BEVERAGE    
9449               1       1.84      4466  AMERICAN VINTAGE BEVERAGE    
9450               1       1.84      4466  AMERICAN VINTAGE BEVERAGE    
9451               1       1.84      4466  AMERICAN VINTAGE BEVERAGE    
9452               1       1.84      4466  AMERICAN VINTAGE BEVERAGE    

[9453 rows x 14 columns]

In [9]:
purchases.groupby(["Brand","PurchasePrice"])[["Quantity","Dollars"]].sum()

Quantity   Dollars
Brand PurchasePrice                    
3140  11.19              4640  51921.60
5215  9.41               4923  46325.43
5255  9.35               6215  58110.25

In [10]:
purchase_prices

Brand                   Description  Price    Size Volume  Classification  \
0   5215  TGI Fridays Long Island Iced  12.99  1750mL   1750               1   
1   5255  TGI Fridays Ultimte Mudslide  12.99  1750mL   1750               1   
2   3140      TGI Fridays Orange Dream  14.99  1750mL   1750               1   

   PurchasePrice  VendorNumber                   VendorName  
0           9.41          4466  AMERICAN VINTAGE BEVERAGE    
1           9.35          4466  AMERICAN VINTAGE BEVERAGE    
2          11.19          4466  AMERICAN VINTAGE BEVERAGE

In [11]:
sales.groupby("Brand")[['SalesQuantity','SalesDollars','SalesPrice','ExciseTax']].sum()

SalesQuantity  SalesDollars  SalesPrice  ExciseTax
Brand                                                    
3140            3890      50531.10    30071.85    7149.25
5215            4651      60416.49    41542.02    8548.96
5255            6096      79187.04    51180.60   11204.28

•The purchases table contains actual purchase data, including the date of purchase, products (brands) purchased by vendors, the amount paid (in dollars), and the quantity purchased.

•The purchase price column is derived from the purchase_prices table, which provides product-wise actual and purchase prices. The combination of vendor and brand is unique in this table.

•The vendor_invoice table aggregates data from the purchases table, summarizing quantity and dollar amounts, along with an additional column for freight.

•This table maintains uniqueness based on vendor and PO number.

•The sales table captures actual sales transactions, detailing the brands purchased by vendors, the quantity sold, the selling price, and the revenue earned.



As the data that we need for analysis is distributed in different tables, we need to create a summary table containing:

purchase transactions made by vendors

sales transaction data

freight(delivery) costs for each vendor

actual product prices from vendors

In [12]:
freight_summary = pd.read_sql_query("""SELECT VendorNumber,SUM(Freight) as FreightCost 
FROM vendor_invoice 
GROUP BY VendorNumber""", conn)
freight_summary

VendorNumber  FreightCost
0               2        27.08
1              54         0.48
2              60       367.52
3             105        62.39
4             200         6.19
..            ...          ...
121         98450       856.02
122         99166       130.09
123        172662       178.34
124        173357       202.50
125        201359         0.09

[126 rows x 2 columns]

p:purchase pp:purchase price

In [13]:
purchase_summary = pd.read_sql_query("""SELECT 
    p.VendorNumber,
    p.VendorName,
    p.Brand,
    p.Description,
    p.PurchasePrice,
    pp.Price as ActualPrice,
    pp.Volume,
    SUM(p.Quantity) AS TotalPurchaseQuantity,
    SUM(p.Dollars) AS TotalPurchaseDollars
FROM purchases p
JOIN purchase_prices pp
    ON p.Brand = pp.Brand
WHERE p.PurchasePrice > 0
GROUP BY p.VendorNumber,p.VendorName, p.Brand, p.Description, p.PurchasePrice, pp.Price, pp.Volume
ORDER BY TotalPurchaseDollars
""",conn)
purchase_summary

VendorNumber                   VendorName  Brand  \
0              7245  PROXIMO SPIRITS INC.          3065   
1              3960  DIAGEO NORTH AMERICA INC      6127   
2              3924  HEAVEN HILL DISTILLERIES      9123   
3              8004  SAZERAC CO INC                5683   
4              9815  WINE GROUP INC                8527   
...             ...                          ...    ...   
10687          3960  DIAGEO NORTH AMERICA INC      3545   
10688          3960  DIAGEO NORTH AMERICA INC      4261   
10689         17035  PERNOD RICARD USA             8068   
10690          4425        MARTIGNETTI COMPANIES   3405   
10691          1128  BROWN-FORMAN CORP             1233   

                       Description  PurchasePrice  ActualPrice Volume  \
0         Three Olives Grape Vodka           0.71         0.99     50   
1      The Club Strawbry Margarita           1.47         1.99    200   
2                  Deep Eddy Vodka           0.74         0.99     50   
3      Dr McGillicuddy's Apple Pie           0.39         0.49     50   
4      Concannon Glen Ellen Wh Zin           1.32         4.99    750   
...                            ...            ...          ...    ...   
10687              Ketel One Vodka          21.89        29.99   1750   
10688       Capt Morgan Spiced Rum          16.17        22.99   1750   
10689             Absolut 80 Proof          18.24        24.99   1750   
10690        Tito's Handmade Vodka          23.19        28.99   1750   
10691      Jack Daniels No 7 Black          26.27        36.99   1750   

       TotalPurchaseQuantity  TotalPurchaseDollars  
0                          1                  0.71  
1                          1                  1.47  
2                          2                  1.48  
3                          6                  2.34  
4                          2                  2.64  
...                      ...                   ...  
10687                 138109            3023206.01  
10688                 201682            3261197.94  
10689                 187407            3418303.68  
10690                 164038            3804041.22  
10691                 145080            3811251.60  

[10692 rows x 9 columns]

In [14]:
sales.columns

Index(['InventoryId', 'Store', 'Brand', 'Description', 'Size', 'SalesQuantity',
       'SalesDollars', 'SalesPrice', 'SalesDate', 'Volume', 'Classification',
       'ExciseTax', 'VendorNo', 'VendorName'],
      dtype='object')

In [18]:
sales_summary = pd.read_sql_query("""SELECT 
    VendorNo,
    Brand,
    SUM(SalesQuantity) AS TotalSalesQuantity,
    SUM(SalesDollars) AS TotalSalesDollars,
    SUM(SalesPrice) AS TotalSalesPrice,
    SUM(ExciseTax) AS TotalExciseTax
    FROM sales
    GROUP BY VendorNo, Brand
    ORDER BY TotalSalesDollars""",conn)
sales_summary

VendorNo  Brand  TotalSalesQuantity  TotalSalesDollars  \
0          8004   5287                   2       9.800000e-01   
1          9206   2773                   1       9.900000e-01   
2          3252   3933                   2       1.980000e+00   
3          3924   9123                   2       1.980000e+00   
4         10050   3623                   2       1.980000e+00   
...         ...    ...                 ...                ...   
11267      3960   3545              135838       4.223108e+06   
11268      3960   4261              200412       4.475973e+06   
11269     17035   8068              187140       4.538121e+06   
11270      4425   3405              160247       4.819073e+06   
11271      1128   1233              142049       5.101920e+06   

       TotalSalesPrice  TotalExciseTax  
0                 0.98            0.10  
1                 0.99            0.05  
2                 0.99            0.10  
3                 0.99            0.10  
4                 1.98            0.10  
...                ...             ...  
11267        545778.28       249587.83  
11268        420050.01       368242.80  
11269        461140.15       343854.07  
11270        561512.37       294438.66  
11271        672819.31       260999.20  

[11272 rows x 6 columns]

In [19]:
# merging all the summary table to create an aggregated summary table for further analysis
vendor_sales_summary = pd.read_sql_query("""WITH FreightSummary AS (
    SELECT 
        VendorNumber, 
        SUM(Freight) AS FreightCost 
    FROM vendor_invoice 
    GROUP BY VendorNumber
), 

PurchaseSummary AS (
    SELECT 
        p.VendorNumber,
        p.VendorName,
        p.Brand,
        p.Description,
        p.PurchasePrice,
        pp.Price AS ActualPrice,
        pp.Volume,
        SUM(p.Quantity) AS TotalPurchaseQuantity,
        SUM(p.Dollars) AS TotalPurchaseDollars
    FROM purchases p
    JOIN purchase_prices pp
        ON p.Brand = pp.Brand
    WHERE p.PurchasePrice > 0
    GROUP BY p.VendorNumber, p.VendorName, p.Brand, p.Description, p.PurchasePrice, pp.Price, pp.Volume
), 

SalesSummary AS (
    SELECT 
        VendorNo,
        Brand,
        SUM(SalesQuantity) AS TotalSalesQuantity,
        SUM(SalesDollars) AS TotalSalesDollars,
        SUM(SalesPrice) AS TotalSalesPrice,
        SUM(ExciseTax) AS TotalExciseTax
    FROM sales
    GROUP BY VendorNo, Brand
) 
SELECT 
    ps.VendorNumber,
    ps.VendorName,
    ps.Brand,
    ps.Description,
    ps.PurchasePrice,
    ps.ActualPrice,
    ps.Volume,
    ps.TotalPurchaseQuantity,
    ps.TotalPurchaseDollars,
    ss.TotalSalesQuantity,
    ss.TotalSalesDollars,
    ss.TotalSalesPrice,
    ss.TotalExciseTax,
    fs.FreightCost
FROM PurchaseSummary ps
LEFT JOIN SalesSummary ss 
    ON ps.VendorNumber = ss.VendorNo 
    AND ps.Brand = ss.Brand
LEFT JOIN FreightSummary fs 
    ON ps.VendorNumber = fs.VendorNumber
ORDER BY ps.TotalPurchaseDollars DESC""",conn)

In [21]:
vendor_sales_summary

VendorNumber                   VendorName  Brand  \
0              1128  BROWN-FORMAN CORP             1233   
1              4425        MARTIGNETTI COMPANIES   3405   
2             17035  PERNOD RICARD USA             8068   
3              3960  DIAGEO NORTH AMERICA INC      4261   
4              3960  DIAGEO NORTH AMERICA INC      3545   
...             ...                          ...    ...   
10687          9815  WINE GROUP INC                8527   
10688          8004  SAZERAC CO INC                5683   
10689          3924  HEAVEN HILL DISTILLERIES      9123   
10690          3960  DIAGEO NORTH AMERICA INC      6127   
10691          7245  PROXIMO SPIRITS INC.          3065   

                       Description  PurchasePrice  ActualPrice Volume  \
0          Jack Daniels No 7 Black          26.27        36.99   1750   
1            Tito's Handmade Vodka          23.19        28.99   1750   
2                 Absolut 80 Proof          18.24        24.99   1750   
3           Capt Morgan Spiced Rum          16.17        22.99   1750   
4                  Ketel One Vodka          21.89        29.99   1750   
...                            ...            ...          ...    ...   
10687  Concannon Glen Ellen Wh Zin           1.32         4.99    750   
10688  Dr McGillicuddy's Apple Pie           0.39         0.49     50   
10689              Deep Eddy Vodka           0.74         0.99     50   
10690  The Club Strawbry Margarita           1.47         1.99    200   
10691     Three Olives Grape Vodka           0.71         0.99     50   

       TotalPurchaseQuantity  TotalPurchaseDollars  TotalSalesQuantity  \
0                     145080            3811251.60            142049.0   
1                     164038            3804041.22            160247.0   
2                     187407            3418303.68            187140.0   
3                     201682            3261197.94            200412.0   
4                     138109            3023206.01            135838.0   
...                      ...                   ...                 ...   
10687                      2                  2.64                 5.0   
10688                      6                  2.34               134.0   
10689                      2                  1.48                 2.0   
10690                      1                  1.47                72.0   
10691                      1                  0.71                86.0   

       TotalSalesDollars  TotalSalesPrice  TotalExciseTax  FreightCost  
0           5.101920e+06        672819.31       260999.20     68601.68  
1           4.819073e+06        561512.37       294438.66    144929.24  
2           4.538121e+06        461140.15       343854.07    123780.22  
3           4.475973e+06        420050.01       368242.80    257032.07  
4           4.223108e+06        545778.28       249587.83    257032.07  
...                  ...              ...             ...          ...  
10687       1.595000e+01            10.96            0.55     27100.41  
10688       6.566000e+01             1.47            7.04     50293.62  
10689       1.980000e+00             0.99            0.10     14069.87  
10690       1.432800e+02            77.61           15.12    257032.07  
10691       8.514000e+01            33.66            4.46     38994.78  

[10692 rows x 14 columns]

This query generates a vendor-wise sales and purchase summary, which is valuable for Performance Optimization:
- The query involves heavy joins and aggregations on large datasets like sales and purchases.
- Helps in analyzing sales, purchases, and pricing for different vendors and brands.
- Future Benefits of Storing this data for faster Dashboarding & Reporting.
- Instead of running expensive queries each time, dashboards can fetch data quickly from vendor_sales_summary.

Let's clean the data if there is any inconsistencies.

# Data Cleaning

In [38]:
vendor_sales_summary.dtypes

VendorNumber               int64
VendorName                object
Brand                      int64
Description               object
PurchasePrice            float64
ActualPrice              float64
Volume                   float64
TotalPurchaseQuantity      int64
TotalPurchaseDollars     float64
TotalSalesQuantity       float64
TotalSalesDollars        float64
TotalSalesPrice          float64
TotalExciseTax           float64
FreightCost              float64
dtype: object

In [23]:
vendor_sales_summary.isnull().sum()

VendorNumber               0
VendorName                 0
Brand                      0
Description                0
PurchasePrice              0
ActualPrice                0
Volume                     0
TotalPurchaseQuantity      0
TotalPurchaseDollars       0
TotalSalesQuantity       178
TotalSalesDollars        178
TotalSalesPrice          178
TotalExciseTax           178
FreightCost                0
dtype: int64

TotalSalesQuantity       178
TotalSalesDollars        178
TotalSalesPrice          178
TotalExciseTax           178 
-> 178 missing values  means vendor has purchased but didnt sell it and also volume is in int but above says object so we got 2 inconsistency.

In [25]:
vendor_sales_summary.head(5)

VendorNumber                   VendorName  Brand              Description  \
0          1128  BROWN-FORMAN CORP             1233  Jack Daniels No 7 Black   
1          4425        MARTIGNETTI COMPANIES   3405    Tito's Handmade Vodka   
2         17035  PERNOD RICARD USA             8068         Absolut 80 Proof   
3          3960  DIAGEO NORTH AMERICA INC      4261   Capt Morgan Spiced Rum   
4          3960  DIAGEO NORTH AMERICA INC      3545          Ketel One Vodka   

   PurchasePrice  ActualPrice Volume  TotalPurchaseQuantity  \
0          26.27        36.99   1750                 145080   
1          23.19        28.99   1750                 164038   
2          18.24        24.99   1750                 187407   
3          16.17        22.99   1750                 201682   
4          21.89        29.99   1750                 138109   

   TotalPurchaseDollars  TotalSalesQuantity  TotalSalesDollars  \
0            3811251.60            142049.0       5.101920e+06   
1            3804041.22            160247.0       4.819073e+06   
2            3418303.68            187140.0       4.538121e+06   
3            3261197.94            200412.0       4.475973e+06   
4            3023206.01            135838.0       4.223108e+06   

   TotalSalesPrice  TotalExciseTax  FreightCost  
0        672819.31       260999.20     68601.68  
1        561512.37       294438.66    144929.24  
2        461140.15       343854.07    123780.22  
3        420050.01       368242.80    257032.07  
4        545778.28       249587.83    257032.07

In [32]:
vendor_sales_summary['VendorName'].unique()

array(['BROWN-FORMAN CORP', 'MARTIGNETTI COMPANIES', 'PERNOD RICARD USA',
       'DIAGEO NORTH AMERICA INC', 'BACARDI USA INC',
       'JIM BEAM BRANDS COMPANY', 'MAJESTIC FINE WINES',
       'ULTRA BEVERAGE COMPANY LLP', 'STOLI GROUP,(USA) LLC',
       'PROXIMO SPIRITS INC.', 'MOET HENNESSY USA INC', 'CAMPARI AMERICA',
       'SAZERAC CO INC', 'CONSTELLATION BRANDS INC', 'M S WALKER INC',
       'SAZERAC NORTH AMERICA INC.', 'PALM BAY INTERNATIONAL INC',
       'REMY COINTREAU USA INC', 'SIDNEY FRANK IMPORTING CO',
       'E & J GALLO WINERY', 'WILLIAM GRANT & SONS INC',
       'HEAVEN HILL DISTILLERIES', 'DISARONNO INTERNATIONAL LLC',
       'EDRINGTON AMERICAS', 'CASTLE BRANDS CORP.',
       'SOUTHERN WINE & SPIRITS NE', 'STE MICHELLE WINE ESTATES',
       'TRINCHERO FAMILY ESTATES', 'MHW LTD', 'WINE GROUP INC',
       'PERFECTA WINES', 'LUXCO INC', 'TREASURY WINE ESTATES',
       'DIAGEO CHATEAU ESTATE WINES', 'SHAW ROSS INT L IMP LTD',
       'PINE STATE TRADING CO', 'PHILLIPS PRO

In [33]:
vendor_sales_summary['Volume'].unique()

array([ 1750. ,   750. ,  1000. ,  1500. ,   375. ,    50. ,  3000. ,
        5000. ,   100. ,   200. ,  4000. ,   187. ,   500. ,   600. ,
         300. ,  1100. ,   250. ,   400. , 18000. ,   150. ,   720. ,
         330. ,   162.5,   180. , 19500. ,  6000. ,   560. , 20000. ,
        9000. ,  3750. ,   650. ])

- The Volume is a numerical column but having object datatype.
- There are some products that are not sold, so having missing values.
- There are white spaces in the categorical columns.

In [34]:
# changing datatype to float
vendor_sales_summary['Volume'] = vendor_sales_summary['Volume'].astype('float')

# filling missing value with 0
vendor_sales_summary.fillna(0,inplace = True)

# removing spaces from categorical columns
vendor_sales_summary['VendorName'] = vendor_sales_summary['VendorName'].str.strip()
vendor_sales_summary['Description'] = vendor_sales_summary['Description'].str.strip()

In [35]:
vendor_sales_summary.isnull().sum()

VendorNumber             0
VendorName               0
Brand                    0
Description              0
PurchasePrice            0
ActualPrice              0
Volume                   0
TotalPurchaseQuantity    0
TotalPurchaseDollars     0
TotalSalesQuantity       0
TotalSalesDollars        0
TotalSalesPrice          0
TotalExciseTax           0
FreightCost              0
dtype: int64

In [39]:
# creating new columns for better analysis
vendor_sales_summary['GrossProfit'] = vendor_sales_summary['TotalSalesDollars'] - vendor_sales_summary['TotalPurchaseDollars']
vendor_sales_summary['ProfitMargin'] = (vendor_sales_summary['GrossProfit'] / vendor_sales_summary['TotalSalesDollars'])*100
vendor_sales_summary['StockTurnover'] = vendor_sales_summary['TotalSalesQuantity'] / vendor_sales_summary['TotalPurchaseQuantity']
vendor_sales_summary['SalesToPurchaseRatio'] = vendor_sales_summary['TotalSalesDollars'] / vendor_sales_summary['TotalPurchaseDollars']

In [40]:
cursor = conn.cursor()

In [41]:
cursor.execute("""CREATE TABLE vendor_sales_summary (
    VendorNumber INT,
    VendorName VARCHAR(100),
    Brand INT,
    Description VARCHAR(100),
    PurchasePrice DECIMAL(10,2),
    ActualPrice DECIMAL(10,2),
    Volume INT,
    TotalPurchaseQuantity INT,
    TotalPurchaseDollars DECIMAL(15,2),
    TotalSalesQuantity INT,
    TotalSalesDollars DECIMAL(15,2),
    TotalSalesPrice DECIMAL(15,2),
    TotalExciseTax DECIMAL(15,2),
    FreightCost DECIMAL(15,2),
    GrossProfit DECIMAL(15,2),
    ProfitMargin DECIMAL(15,2),
    StockTurnover DECIMAL(15,2),
    SalesToPurchaseRatio DECIMAL(15,2),
    PRIMARY KEY (VendorNumber, Brand)
);
""")

In [44]:
pd.read_sql_query("select * from vendor_sales_summary",conn)

VendorNumber                VendorName  Brand  \
0              1128         BROWN-FORMAN CORP   1233   
1              4425     MARTIGNETTI COMPANIES   3405   
2             17035         PERNOD RICARD USA   8068   
3              3960  DIAGEO NORTH AMERICA INC   4261   
4              3960  DIAGEO NORTH AMERICA INC   3545   
...             ...                       ...    ...   
10687          9815            WINE GROUP INC   8527   
10688          8004            SAZERAC CO INC   5683   
10689          3924  HEAVEN HILL DISTILLERIES   9123   
10690          3960  DIAGEO NORTH AMERICA INC   6127   
10691          7245      PROXIMO SPIRITS INC.   3065   

                       Description  PurchasePrice  ActualPrice  Volume  \
0          Jack Daniels No 7 Black          26.27        36.99  1750.0   
1            Tito's Handmade Vodka          23.19        28.99  1750.0   
2                 Absolut 80 Proof          18.24        24.99  1750.0   
3           Capt Morgan Spiced Rum          16.17        22.99  1750.0   
4                  Ketel One Vodka          21.89        29.99  1750.0   
...                            ...            ...          ...     ...   
10687  Concannon Glen Ellen Wh Zin           1.32         4.99   750.0   
10688  Dr McGillicuddy's Apple Pie           0.39         0.49    50.0   
10689              Deep Eddy Vodka           0.74         0.99    50.0   
10690  The Club Strawbry Margarita           1.47         1.99   200.0   
10691     Three Olives Grape Vodka           0.71         0.99    50.0   

       TotalPurchaseQuantity  TotalPurchaseDollars  TotalSalesQuantity  \
0                     145080            3811251.60            142049.0   
1                     164038            3804041.22            160247.0   
2                     187407            3418303.68            187140.0   
3                     201682            3261197.94            200412.0   
4                     138109            3023206.01            135838.0   
...                      ...                   ...                 ...   
10687                      2                  2.64                 5.0   
10688                      6                  2.34               134.0   
10689                      2                  1.48                 2.0   
10690                      1                  1.47                72.0   
10691                      1                  0.71                86.0   

       TotalSalesDollars  TotalSalesPrice  TotalExciseTax  FreightCost  \
0           5.101920e+06        672819.31       260999.20     68601.68   
1           4.819073e+06        561512.37       294438.66    144929.24   
2           4.538121e+06        461140.15       343854.07    123780.22   
3           4.475973e+06        420050.01       368242.80    257032.07   
4           4.223108e+06        545778.28       249587.83    257032.07   
...                  ...              ...             ...          ...   
10687       1.595000e+01            10.96            0.55     27100.41   
10688       6.566000e+01             1.47            7.04     50293.62   
10689       1.980000e+00             0.99            0.10     14069.87   
10690       1.432800e+02            77.61           15.12    257032.07   
10691       8.514000e+01            33.66            4.46     38994.78   

       GrossProfit  ProfitMargin  StockTurnover  SalesToPurchaseRatio  
0       1290667.91     25.297693       0.979108              1.338647  
1       1015032.27     21.062810       0.976890              1.266830  
2       1119816.92     24.675786       0.998575              1.327594  
3       1214774.94     27.139908       0.993703              1.372493  
4       1199901.61     28.412764       0.983556              1.396897  
...            ...           ...            ...                   ...  
10687        13.31     83.448276       2.500000              6.041667  
10688        63.32     96.436186      22.333333             28.059829  
10689         0.50 

In [43]:
vendor_sales_summary.to_sql('vendor_sales_summary', conn, if_exists = 'replace', index = False )

10692

In [2]:
import sqlite3
import pandas as pd
import logging
import os
from ingestion_db import ingest_db

# Ensure logs folder exists
os.makedirs("logs", exist_ok=True)

# 🧹 Clear all existing handlers (important in Jupyter!)
for handler in logging.root.handlers[:]:
    logging.root.removeHandler(handler)
logging.basicConfig(
    filename="logs/get_vendor_summary.log", 
    level=logging.DEBUG,
    format="%(asctime)s - %(levelname)s - %(message)s", 
    filemode="a"  
)

def ingest_db(df, table_name, engine):
    '''this function will ingest the dataframe into database table'''
    df.to_sql(table_name, con = engine, if_exists = 'replace', index = False)
    
def create_vendor_summary(conn):
    '''this function will merge the different tables to get the overall vendor summary and adding new columns in the resultant data'''
    vendor_sales_summary = pd.read_sql_query("""WITH FreightSummary AS (
        SELECT 
            VendorNumber, 
            SUM(Freight) AS FreightCost 
        FROM vendor_invoice 
        GROUP BY VendorNumber
    ), 
    
    PurchaseSummary AS (
        SELECT 
            p.VendorNumber,
            p.VendorName,
            p.Brand,
            p.Description,
            p.PurchasePrice,
            pp.Price AS ActualPrice,
            pp.Volume,
            SUM(p.Quantity) AS TotalPurchaseQuantity,
            SUM(p.Dollars) AS TotalPurchaseDollars
        FROM purchases p
        JOIN purchase_prices pp
            ON p.Brand = pp.Brand
        WHERE p.PurchasePrice > 0
        GROUP BY p.VendorNumber, p.VendorName, p.Brand, p.Description, p.PurchasePrice, pp.Price, pp.Volume
    ), 
    
    SalesSummary AS (
        SELECT 
            VendorNo,
            Brand,
            SUM(SalesQuantity) AS TotalSalesQuantity,
            SUM(SalesDollars) AS TotalSalesDollars,
            SUM(SalesPrice) AS TotalSalesPrice,
            SUM(ExciseTax) AS TotalExciseTax
        FROM sales
        GROUP BY VendorNo, Brand
    ) 
    
    SELECT 
        ps.VendorNumber,
        ps.VendorName,
        ps.Brand,
        ps.Description,
        ps.PurchasePrice,
        ps.ActualPrice,
        ps.Volume,
        ps.TotalPurchaseQuantity,
        ps.TotalPurchaseDollars,
        ss.TotalSalesQuantity,
        ss.TotalSalesDollars,
        ss.TotalSalesPrice,
        ss.TotalExciseTax,
        fs.FreightCost
    FROM PurchaseSummary ps
    LEFT JOIN SalesSummary ss 
        ON ps.VendorNumber = ss.VendorNo 
        AND ps.Brand = ss.Brand
    LEFT JOIN FreightSummary fs 
        ON ps.VendorNumber = fs.VendorNumber
    ORDER BY ps.TotalPurchaseDollars DESC""",conn)

    return vendor_sales_summary


def clean_data(df):
    '''this function will clean the data'''
    # changing datatype to float
    df['Volume'] = df['Volume'].astype('float')
    
    # filling missing value with 0
    df.fillna(0,inplace = True)
    
    # removing spaces from categorical columns
    df['VendorName'] = df['VendorName'].str.strip()
    df['Description'] = df['Description'].str.strip()

    # creating new columns for better analysis
    df['GrossProfit'] = df['TotalSalesDollars'] - df['TotalPurchaseDollars']
    df['ProfitMargin'] = (df['GrossProfit'] / df['TotalSalesDollars'])*100
    df['StockTurnover'] = df['TotalSalesQuantity'] / df['TotalPurchaseQuantity']
    df['SalesToPurchaseRatio'] = df['TotalSalesDollars'] / df['TotalPurchaseDollars']

    
    return df

if __name__ == '__main__':
    # creating database connection
    conn = sqlite3.connect('inventory.db')
    
    logging.info('Creating Vendor Summary Table.....')
    summary_df = create_vendor_summary(conn)
    logging.info(summary_df.head())
    
    logging.info('Cleaning Data.....')
    clean_df = clean_data(summary_df)
    logging.info(clean_df.head())
    
    logging.info('Ingesting data.....')
    ingest_db(clean_df,'vendor_sales_summary',conn)
    logging.info('Completed')